### ISSUE TITLE
Ordered comparison with pointer not detected

### ISSUE THREAD

**Issue by d9f35d94-5ded-4402-8ddf-8e60f27bb643** [llvm/llvm-project#14107]
|  |  |
| --- | --- |
| Bugzilla Link | [13735](https://llvm.org/bz13735) |
| Version | 3.0 |
| OS | All |
| CC | @belkadan,@tkremenek |

## Extended Description 
```
--------------------------
int main (void)
{
  void *volatile qwe = main;

  if (qwe < 0)
    return 1;

  return 0;
}
--------------------------
$ gcc -Wall -Wextra zxc.c
zxc.c: In function 'main':
zxc.c:5:11: warning: ordered comparison of pointer with integer zero [-Wextra]
zxc.c:8:11: warning: comparison between pointer and integer [enabled by default]
$ clang -Wall -Wextra --analyze zxc.c 
$ gcc --version
gcc-4.6.real (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3
$ clang --version
Ubuntu clang version 3.0-6ubuntu3 (tags/RELEASE_30/final) (based on LLVM 3.0)
--------------------------
```
So, Clang does not detect very important bug in my code.

**Comment by Quuxplusone** [comment]
Looks like this has been detected by `-Wpedantic` since Clang 3.0.0:
```
<source>:5:11: warning: ordered comparison between pointer and zero ('void *' and 'int') is an extension [-pedantic]
  if (qwe < 0)
      ~~~ ^ ~
```
It's a hard error in C++, for what that's worth.
